src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,407 @@
+            String status = container.getStatus();
+            
+            if (status == null) {
+                LOGGER.warn("Container {} has a null-status and thus cannot be checked; ignoring container", containerId);
+                continue;
+            }
+            
+            
+            Map<String, String> containerLabels = getLabelsOfContainer(client, containerId);
+            
+            String containerNodeName = containerLabels.get(DockerTemplate.CONTAINER_LABEL_NODE_NAME);
+            if (containerNodeName == null) {
+                LOGGER.warn("Container {} is said to be created by this Jenkins instance, but does not have any node name label assigned; manual cleanup is required for this", containerId);
+                continue;
+            }
+            
+            result.registerMapping(container, containerNodeName);
+        }
+        
+        return result;
+    }
+
+    private static Map<String, String> getLabelsOfContainer(DockerClient client, String containerId) {
+        // TODO may be called multiple times => may be cached to save performance
+        InspectContainerResponse icr = client.inspectContainerCmd(containerId).exec();
+        return icr.getConfig().getLabels();
+    }
+
+    private void cleanupSuperfluousContainers(DockerClient client, ContainerNodeNameMapping csm, DockerCloud dc) {
+        Collection<Container> allContainers = csm.getAllContainers();
+        
+        for (Container container : allContainers) {
+            String nodeName = csm.getNodeName(container.getId());
+            
+            Node node = this.nodeMap.get(nodeName);
+            if (node != null) {
+                // the slave and the container still have a proper mapping => ok
+                continue;
+            }
+            
+            /*
+             * During startup it may happen temporarily that a container exists, but the
+             * corresponding node isn't there yet.
+             * That is why we have to have a grace period for pulling up containers.
+             */
+            if (this.isStillTooYoung(container.getCreated()))
+                continue;
+
+            // this is a container, which is missing a corresponding node with us
+            LOGGER.info("Container {}, which is reported to be assigned to node {}, is no longer associated (slave might be gone already?)", container.getId(), nodeName);
+            LOGGER.info("Container {}'s last status is {}; it was created on {}", container.getId(), container.getStatus(), container.getCreated());
+            
+            try {
+                this.terminateContainer(dc, container);
+            } catch (Exception e) {
+                // Graceful termination failed; we need to use some force
+                LOGGER.warn("Graceful termination of Container {} failed; terminating directly via API - this may cause remnants to be left behind", container.getId(), e);
+                client.removeContainerCmd(container.getId());
```

Comment:
> I think that we'd be better off removing this call to `client.removeContainerCmd` entirely 

The meaning of your statement may be ambiguous to me. It could mean that you *never* want `client.removeContainerCmd` to have called, or that you just want to have the call outsourced in an own `private` method.  With 97e119440c8be618092b9fd73463a652815c65b2 I now did the latter.

PS: I also don't like the pure `removeContainerCmd` case here either. But I came to this point, when I imagined the following use case: Someone has configured the config of `DockerCloudTemplate` totally wrong, which causes containers to be pulled up in masses, but they can't be attached as node (e.g. the password for the SSH credentials is wrong). To get rid of it, the administrator deletes the `DockerCloudTemplate` configuration. Yet, there are dozens of containers vagabonding around on the `DockerCloud`. As we don't have a `DockerCloudTemplate` anymore, we cannot delete them anymore, as we cannot create a `DockerTransientNode` anymore. Still, those containers have our JenkinsId in the labels...